Add cookbook and model co-design refs (#22)
stas00/ml-engineering: performance/software.md

If your network is very slow, your training is likely to be network-bound and many improvements in the training setup will not help with the improving performance.

> Note: The [cookbook](https://github.com/EleutherAI/cookbook) contains a set of [communication benchmarks](https://github.com/EleutherAI/cookbook/tree/main/benchmarks/communication) for each collective that you can use to quickly measure the throughput of your internode or intranode network.

Reviewer: can we move it to https://github.com/stas00/ml-engineering/tree/master/network 

I need to do a lot more massaging to make everything well organized.

Author: will move

Reviewer: Hmm, I will need to rework things first, it's perfect where you placed it in the first place until then. So nothing to change.

## Vector and matrix size divisibility

> A new paper, [The Case for Co-Designing Model Architectures with Hardware](https://arxiv.org/abs/2401.14489) investigates the effects of transformer sizing on the underlying hardware. The [associated scripts](https://github.com/EleutherAI/cookbook/tree/main/benchmarks/sizing) allow you to run the benchmarks yourself if you're running on hardware besides NVIDIA V100/A100.

Reviewer: ```suggestion
The paper, [The Case for Co-Designing Model Architectures with Hardware](https://arxiv.org/abs/2401.14489) investigates the effects of transformer sizing on the underlying hardware. The [associated scripts](https://github.com/EleutherAI/cookbook/tree/main/benchmarks/sizing) allow you to run the benchmarks yourself if you're running on hardware besides NVIDIA V100/A100.
```

This is not a blog and I'm lazy so let's future proof it as it won't be new in 6 months.

For parameters that are small, there is also [Dimension Quantization Effects](https://docs.nvidia.com/deeplearning/performance/dl-performance-matrix-multiplication/index.html#dim-quantization) to consider, this is where tiling happens and the right multiplier can have a significant speedup.

> The work in https://arxiv.org/abs/2401.14489 provides much greater detail on tile/wave quantization and the number of attention heads, but the highlights are:

Reviewer: ```suggestion
[The Case for Co-Designing Model Architectures with Hardware](https://arxiv.org/abs/2401.14489) provides much greater detail on tile/wave quantization and the number of attention heads, but the highlights are:
```

### Tile and wave quantization

> <img src="images/notation.png" width=20% height=20% />

Reviewer: These images render too tiny in the markdown, at least I couldn't read them

Should we leave them as they are, as this is not a paper and there is no need to cram images into tiny columns as it's done in papers, which can't be read.

Let's just use?
```suggestion
![notation](images/notation.png)
```

Reviewer: I can just go over those and make them a bit smaller if they are indeed too big. I am likely to have time tomorrow night to adjust those, I will take care of those. You did the heavy lifting already, Quentin.

First, some background.

NVIDIA GPUs divide the output matrix into regions or tiles as shown in the below figure and schedule them to one of the available streaming multiprocessors (SM) on the GPU (e.g., A100 GPUs have 108 SMs). Each tile or thread block is processed in a Tensor Core, which NVIDIA introduced for fast tensor operations. NVIDIA Tensor Cores are only available for GEMMs with appropriate dimensions. Tensor Cores can be fully utilized when GEMM dimensions $m$, $k$, and $n$ are multiples of 16 bytes and 128 bytes for V100 and A100 GPUs, respectively. Since a FP16 element is 2 bytes, this corresponds to dimension sizes that are multiples of 8 and 64 elements, respectively. If these dimension sizes are not possible, Tensor Cores perform better with larger multiples of 2 bytes. 

<img src="images/tiling.png" width=15% height=15% />

> There are multiple tile sizes that the kernel can choose from. If the GEMM size does not divide evenly into the tile size, there will be wasted compute, where the thread block must execute fully on the SM, but only part of the output is necessary. This is called the tile quantization effect, as the output is quantized into discrete tiles. 

Reviewer: ```suggestion
There are multiple tile sizes that the kernel can choose from. If the GEMM size does not divide evenly into the tile size, there will be wasted compute, where the thread block must execute fully on the SM, but only part of the output is necessary. This is called the **tile quantization** effect, as the output is quantized into discrete tiles. 
```

> Another quantization effect is called wave quantization. As the thread blocks are scheduled to SMs, only 108 thread blocks at a time may be scheduled. If, for example, 109 thread blocks must be scheduled, two rounds, or waves, of thread blocks must be scheduled to GPU. The first wave will have 108 thread blocks, and the second wave will have 1. The second wave will have almost the same latency as the first, but with a small fraction of the useful compute. As the matrix size increases, the last or tail wave grows. The throughput will increase, until a new wave is required. Then, the throughput will drop.

Reviewer: ```suggestion
Another quantization effect is called **wave quantization**. As the thread blocks are scheduled to SMs, only 108 thread blocks at a time may be scheduled. If, for example, 109 thread blocks must be scheduled, two rounds, or waves, of thread blocks must be scheduled to GPU. The first wave will have 108 thread blocks, and the second wave will have 1. The second wave will have almost the same latency as the first, but with a small fraction of the useful compute. As the matrix size increases, the last or tail wave grows. The throughput will increase, until a new wave is required. Then, the throughput will drop.
```

What this means for transformers, is that for a given ratio of $h/a$, one needs to ensure they're on the crest of a wave. If you're using NVIDIA V100/A100 GPUs, we've already done this work for you in https://arxiv.org/pdf/2401.14489.pdf

An example of this for 32 attention heads:

<img src="images/wave-quant.png" width=30% height=30% />

More powers of 2 in $h/a$ helps! 

### Number/size of Attention heads

> Generally, it's most computationally efficient to keep the ratio of $h/a$ as large as possible without accuracy degradation. A good figure from https://arxiv.org/pdf/2401.14489.pdf showing this effect is:

Reviewer: ```suggestion
Generally, it's most computationally efficient to keep the ratio of $h/a$ as large as possible without accuracy degradation. A good figure from [The Case for Co-Designing Model Architectures with Hardware](https://arxiv.org/abs/2401.14489) showing this effect is:
```
